qutebrowser/completion/models/sqlmodel.py

```diff
@@ -0,0 +1,217 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+# Copyright 2016 Ryan Roden-Corrent (rcorre) <ryan@rcorre.net>
+#
+# This file is part of qutebrowser.
+#
+# qutebrowser is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""A completion model backed by SQL tables."""
+
+import re
+
+from PyQt5.QtCore import Qt, QModelIndex, QAbstractItemModel
+from PyQt5.QtSql import QSqlTableModel, QSqlDatabase, QSqlQuery
+
+from qutebrowser.utils import usertypes, log
+
+
+class SqlCompletionModel(QAbstractItemModel):
```

Comment:
There's no 'natural' way to make a `SqlTableModel`/`SqlQueryModel` look like a tree. This is a wrapper around several `SqlTableModels`, each of which populates a single category. I hope I'm not losing out on lazy loading here! At the very least I'm still relying on `SqlTableModel` for filtering!